### ISSUE TITLE
Alerting support for queries using template variables

### ISSUE THREAD

**Issue by calind** [grafana/grafana#6557]
It would be pretty useful if grafana would support alerting for queries using template variables. The way I see it work it would be as follows:

1. Generate queries foreach template variable combination (discarding template variable for __all__) 
2. When generating queries, consider the frozen list if the template variable is set to never refresh, else update the template variable list
3. Allow filtering (trough regex or by providing a static value) for each template variable

The current workaround is to use an invisible wildcard metric, but the problem I see with this approach is that it loses context.

**Comment by oiooj** [comment]
+1

**Comment by bergquist** [comment]
1. What would be the difference compared to just using all?

**Comment by antoiner77** [comment]
+1
Would be nice to be able to add alerting on server with a low life time (AWS auto scaling), auto register the server on grafana is easy with the templating but it's sad to not be able to put alerting on them

**Comment by calind** [comment]
@bergquist it's unpractical using all for example when you have more than a dozen hosts.

![nivex6impyskjxkpmldv](https://cloud.githubusercontent.com/assets/266375/20298310/d885643c-ab1d-11e6-944c-9475a613b9bb.png)

If for example only few of them are failing, (let's say 5), it is very useful to receive an email for each failing alert. This way is also much easier to integrate with other tools which in general expect one alert per metric.

The current approach (using all) is pretty neat though when there are fewer instances or when you are alerting at service level (eg. # of jobs in queue).

**Comment by Deshke** [comment]
what @calind said, i've got multiple $host variables wich are working fine with the influxDB but not with the alerts

**Comment by NotSoCleverLogin** [comment]
+1 as well.

Just a thought, since you are able to query with a template variable, wouldn't you just be able to do the same query with the alerting metrics and maybe iterate through the results to see which meet the alert criteria?

**Comment by bergquist** [comment]
@NotSoCleverLogin It would be possible. But would you want to change the behavior of alert rule based on what template varlue are selected? 

Using the all option for the template is the only way that makes sense for me.

**Comment by mstaalesen** [comment]
+1

I have a setup of X environments with the same components in each environment. We are currently using prometheus to alert on e.g cpu usage/disk usage etc. There we specify an alert for a query, and when the alert is triggered it will just state which environment the alert was triggered from. 

If we would do this with the All variable, that would work to some extent. But, using @calind's example, the screenshot would be filled with the trend of all cpus from all of my environments, and not just the environment where I would want to be informed about said problem. The graph will (or can) be obscured with information from other environments. In some scenarios it could be interesting to compare cpu in other environments, but there are no guarantees that what is happening in a test environment is happening in our production environment, etc.  

We are also looking into creating dashboards that can be used by operations, showing annotations for alerts in the "standard" overview dashboard. Given that we use 'env' template variables for these kind of dashboards it's not really possible for us to do that with how it is implemented right now. I would have to manually (at least to some extent) generate a "shadow" dashboard where the alerts are triggered (which makes me loose the annotations in the overview dashboard).

Another thing I think template variables can help you do is to route the alerts (should you choose to implement such a feature) to different sources (some to operations if in production, to qa/developers if in test environments etc).

**Comment by StianOvrevage** [comment]
+1 for supporting alerts on templated queries.

**Comment by calind** [comment]
@bergquist, some dashboards don't have an _All_ option. For example system metrics by collectd (https://grafana.net/dashboards/24). Having an _All_ option would certainly no be practical for let's say 10 or more servers. That's why the need to iterate trough template variables.

**Comment by StianOvrevage** [comment]
Allowing use of All is a good and welcomed start.

In Prometheus, queries need to be written in a different way to allow All:

    some.metric{hostname=~"$Hostname"}

Notice the extra tilde there, allowing for regular expression searching (and the wildcard in All).

I have not benchmarked the possible performance impact of going from a straight query to a regex search query but at least for now it would apparently solve our problems.

**Comment by max3163** [comment]
+1

**Comment by steverweber** [comment]
+1
not sure how it should be implemented, just know it's needed.

**Comment by Krylon360** [comment]
+1 
We use Prometheus as the Datasource to monitor our Kubernetes Infrastructure for bout our On-Prem K8S Clusters and our AWS K8S Clusters.
All of our dashboards use Templated Variables for the Datasource ($Environment), $Instance/Node, $Namespace, and $Pod.
Due to the way the Prometheus Query Structure is; all of the queries have Templated Variables; which prevents the Alert Rules from allowing to save.
I would love to see Templated Variable Queries added to the alerting.

**Comment by andrewawagner** [comment]
+1

**Comment by shervinkh** [comment]
+1
We use templating dashboards for multi-server environment which is the logical way (and many people use), So we can't use alerting with grafana right now. The only way is to have a separate non-templating dashboard or setup alerting with prometheus itself which is not easy.

**Comment by steverweber** [comment]
perhaps if there was an option or simple way to save/export a dashboard with the template variables backed into all the fields... this would perhaps be a good half way point until another solution is found.

**Comment by daraeburn** [comment]
+1 for supporting alerts on templated queries. We currently use templating on all our dashboards so can't take advantage of this really cool feature.

**Comment by tsn77130** [comment]
+1, we have a lot of templated dashboards, and we can't use alerting for now, we have to deducplicate dashboards for having alerts, and we so lose templating power

**Comment by drewboswell** [comment]
+1, Almost all of our dashboards use templated variables (and nested templated variables). 

We would like to be able to set alerts on repeat panels to get individual alerts per template-variable group if needed. Plus this means that the alerting is dynamic and not super manual as it is now.

**Comment by ebirukov** [comment]
+1

**Comment by erSitzt** [comment]
When working with grafana it feels like templating is encouraged everywhere and it feels wrong to create an extra set of graphs not using variables just to use the alerting feature...

**Comment by kanwangzjm** [comment]
+1 for supporting alerts on templated queries.
also, we found that when we use Chinese ruleName or Chinese title, we received abnormal email with rule triggered. For example, we expected “个股分时线接口请求时间(getTimeTrend) alert” but received "ä¸ªè¡åæ¶çº¿æ¥å£è¯·æ±æ¶é´(getTimeTrend) alert", maybe the charset  is not correct.

**Comment by AlexMaksimkin** [comment]
+1 to implement templated vars in alerts

**Comment by fingul** [comment]
+1

**Comment by drew-royster** [comment]
+1 would get a great addition

**Comment by tj13** [comment]
+1

**Comment by redhootcp** [comment]
+1 to implement templated vars in alerts

**Comment by bbae-dev** [comment]
+1

**Comment by bbae-dev** [comment]
+1 looking forward for it

**Comment by clhlc** [comment]
+1

**Comment by staslev** [comment]
+1

**Comment by actionjax** [comment]
+1

**Comment by jesseorr** [comment]
+1

**Comment by thetechnick** [comment]
Please stop writing `+1`!
Everybody that has subscribed to this issue will get an email...

There is a github feature only to get rid of those `+1` comments:
https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments

**Comment by StianOvrevage** [comment]
@thetechnick There is a link in the e-mail where you can mute the thread and not receive any e-mails. But I understand that you might want to just get notified when the feature is complete, but I also like to get the issue bumped so that it hopefully will get worked on sooner :)

**Comment by tomekit** [comment]
I am missing this feature as well.

On top of that there might be a bug in a way Grafana detect whether metric used in `query` uses the template variables.

When you've a series which uses the template variables indirectly, Grafana does not stop you to add that series as an alert. The alert obviously does not work correctly.

See the `#K` (it uses #D, which uses #A and #A uses templ. var):
![grafana](https://cloud.githubusercontent.com/assets/7759113/21602171/6052a520-d191-11e6-9f15-f46337900fa0.png)

I could still select it:
![grafana2](https://cloud.githubusercontent.com/assets/7759113/21602173/64ef766c-d191-11e6-86c1-a2f48004be76.png)

**Comment by marketadvorackova** [comment]
Templates everywhere, which means alerting no where. 
Not sure how the alerting has been implemented, but for a simple graph the query gets "translated", template variables substituted with values, before making a call to the data source, right? So why not in this case? In any way, as said before, having almost all of the queries using template variables, alerting is completely of for me. Please, could you implement it so that we don't have to move alerting outside Grafana? Thanks a lot!

**Comment by erSitzt** [comment]
I think we should recognise that alerting with templating is not trivial and i think the ALL options is the way to go because we dont want our alerts changing when someone is using the dashboard.
But grafana still would have to create new alerts if the template query returns new results... which happens quiet often as we scale our apps.
This leads to more problems if you are using InfluxDB as many of us are using tags/tag values i guess, and there is no time filter for them... so grafana would create alerts for all service that ever existed on any host...

**Comment by marketadvorackova** [comment]
ALL is an option, sure. More flexible would be a recognition of the template variables in the query and letting the user set the values up in the alert condition configuration. The best, but complicated I guess, would be to have multiple alerts (the same way there are multiple queries) so that a different alert could be set up for a different template variable values in the query. This would enable the administrator to set up different alert conditions for different hosts for example.

**Comment by pdf** [comment]
Multiple alerting profiles would be great, but for an initial pass, just providing the same template selectors as are available on the dashboard in the alerting panel would solve a lot of problems.

I think there should be an toggle for each variable to aggregate results for that variable into a single notification.  This provides a simple but effective method to control the verbosity of notifications - you may want to notify only once for multiple related metrics, but notify for each host where any metric is failing.  Or, you may want to notify only once for a failing metric no matter how many hosts are affected.

**Comment by siteshbehera** [comment]
do we have any targeted milestone for this bug ?

**Comment by tomekit** [comment]
I had some issues with the alerting on a complicated queries as well as queries that uses the template variables. I've found out easy workaround, which maybe not pretty, but it works for my use case.
It's just extracting the query after you built it, so there are no template variables and any #ROW references. This could be obvious for you, there is no rocket science, but to me it was life changer.

What I do is I prepare a query:
![image](https://cloud.githubusercontent.com/assets/7759113/21962175/a4f4cf18-db1f-11e6-9745-bcb0fefc17c3.png)

then extract it using the Chrome dev tools:
![image](https://cloud.githubusercontent.com/assets/7759113/21962172/94b95f92-db1f-11e6-9041-17fb03f3a1c9.png)

Put it in another row (toggle edit mode):
![image](https://cloud.githubusercontent.com/assets/7759113/21962178/b999e71e-db1f-11e6-89cf-c6d40628c5a8.png)

Set up the alerting:
![image](https://cloud.githubusercontent.com/assets/7759113/21962180/c567a7ca-db1f-11e6-90c8-350adb937e0c.png)

Voila !

**Comment by bergquist** [comment]
@siteshbehera This is not a bug. Its a feature request.

But no. We dont have a milestone for this currently.

**Comment by lastsky** [comment]
artificial intelligence grafana plugin should be included in commit for this feature.

**Comment by 1stdemon** [comment]
Waiting for templates in Alerts too +1

**Comment by Faradax** [comment]
I'm also very much in favor of what calind provided as possible implementation in the opening post. It seems to fits neatly into how many (me included) use templated dashboards - where you have one dashboard, but switch/limit some variables to manually look at specific things. I think the example of the "server"-variable might be the most fitting one. There, the template variable (without _all_-value) would become something not unlike a "_tab_" in my dashboard - I can switch between them to see different sets of data. It's then easy to assume that, when setting up an alert, the alert would exist for each possible "_tab_" seperately.

**Comment by fahimeh2010** [comment]
Waiting to support templates in Alerts +1

**Comment by danhallin** [comment]
As a previous user of Librato where alerting was partially capable of this templating I'd like to chime in with an equally partial solution. In Librato each metric comes with a 'source' variable, and alerts on a graph would automatically be per source.

I think an equal solution would satisfy the needs raised here. When creating an alert you should be able to choose a single template value as a 'source' and this source is mentioned in the alert, all others being set to 'all'. This solution at least avoids the combinatorics problem you get if allowing the use of multiple template variables.

Myself I just set an invisible max or min graph over the data I'm interested and do the alert on it, not as powerful but still a working solution until this issue is resolved.

**Comment by IldarMinaev** [comment]
+1

**Comment by dennisosimon** [comment]
+1

**Comment by nadirollo** [comment]
Hi, I'm definitely looking for this feature as, with the majority of cases before, all of my dashboards are using templated queries to support multiple environments (at least).

Is there a place where the roadmap of grafana is being tracked? Or any way that we can see if features (like this one) will be implemented in a future (near or not so near) without poking mantainers on github? :)

**Comment by yannispanousis** [comment]
Amazing really eagerly waiting for this one

**Comment by josephzilberman** [comment]
+1

**Comment by AlexeyKolychev** [comment]
+1

**Comment by bergquist** [comment]
We are still not sure how to do this.

I think reusing the selected template variable for alerting would be dangerous since people can choose to view just one option and then forget to change back to `All` or something wider. I would not feel safe with such behavior. Alert rules should be extremely easy to understand and reason about. Explicit rules > magic rules.

One solution for this problem would be to have two values for each template variable. One for visualization in the dashboard and one for alerting. This would make it possible to always have the wider option in alerting and still select just a few options in the graphs. Connecting those values should be possible but not the default behavior. 

That solution how every would be fairly large and complex feature.

**Comment by calind** [comment]
I have two proposals for a solution.

1. A short term proposal is to add an alerting option so that in the rendered garph (the one being sent by email) only metrics which are alerting are being displayed. This would solve the clutter when the alerting graph contains dozens of metrics.

2. A long term solution would be to iterate trough template variables, so that you have an distinct alert for each template value combination.

**Comment by StianOvrevage** [comment]
As I mentioned back in november. For prometheus users just using 'All' as variable value is sufficient if the queries are written properly ( some.metric{hostname=~"$Hostname"} ). 

Should probably also be very easy to implement.

**Comment by pdf** [comment]
@bergquist I think option 2 is heading in the right direction (similar to what I suggested in https://github.com/grafana/grafana/issues/6557#issuecomment-272588490), it doesn't seem too complex, since the code to handle template var selection already exists for the dashboard, and there's no need to duplicate the var configuration, just selection.  I don't think I'd bother connecting the dashboard selection to alerting in the first pass at this feature.

**Comment by rahulredde9** [comment]
i solved it by creating a new metric query just for the alert, without the template variables and disabled it (to exclude it from the graph) on Grafana version v4.1.1.

**Comment by arnobroekhof** [comment]
+1

**Comment by seb-margineanu-ft** [comment]
+1 to implement templated vars in alerts

**Comment by alejandroandreu** [comment]
Does this affect _all_ versions of Grafana? Or was this a feature that was available before? This is kind of a deal breaker for me and wouldn't mind installing a previous version.

**Comment by gjgreen** [comment]
@alejandroandreu alerting was added in version 4, it has never worked with templating.

**Comment by iter00** [comment]
+1 to implement templated vars in alerts

**Comment by shanielh** [comment]
I would like to be able to select / enter the combinations that the alerts should evaluate, since some of the environments that I run are not production environments, There are two ways to implement this, The first one is more explicit, the second one is easier to configure.

##### Enter all the wanted combinations manually

* This configuration should be shown in the Alert configuration panel

For example, If I have 3 templates variables : `cloud`, `region` and `type`, I would fill a table that looks like this : 

| cloud | region     | type |
|-------|------------|------|
| aws   | us-east-1  | prod |
| aws   | us-west-1  | prod |
| azure | Central US | prod |

The table should have extra row for inserting new rows and delete button for every row.

##### Enter possible values and Grafana will calculate Cartesian Product

* Note: This configuration can be entered in the template variable configuration panel

| cloud | region     | type |
|-------|------------|------|
| aws   | us-east-1  | prod |
| azure | us-west-1  |      |
|       | Central US |      |

The combinations that will be created for this input are : 

 - `aws` `us-east-1` `prod`
 - `aws` `us-west-1` `prod`
 - `aws` `Central US` `prod`
 - `azure` `us-east-1` `prod`
 - `azure` `us-west-1` `prod`
 - `azure` `Central US` `prod`

*But* Grafana can handle this situation by "entering" the first variable (`cloud`), and then filtering the available values from the second variable (`region`) until it finds all the possible combinations (note - it should be done iteratively for all the variables) . This is possible when people use queries in the tags like this : 

```sql
SHOW TAG VALUES WITH KEY = "REGION" WHERE "CLOUD" =~ /$CLOUD/	
```

And in this case, the produced combinations will be fine (which is the same as the table in the first option) : 

 - `aws` `us-east-1` `prod`
 - `aws` `us-west-1` `prod`
 - `azure` `Central US` `prod`


I hope my suggestions will be helpful.

**Comment by dbcook** [comment]
We have this issue (OpenTSDB data source) in a slightly different context - if you use a template var to select a downsample interval in the metric, the alert query fails with error 400.  I understand the difficulties in implementing a general solution but we are going to have to re-engineer various existing dashboards to enable alerting.

**Comment by pdf** [comment]
@dbcook sounds like a distinct problem that you should probably file a separate issue for.

**Comment by skygragon** [comment]
Templating is really an awesome feature, and so is alerting. We'd better to have them working together smoothly instead of any awkward workaround.

**Comment by mathurj** [comment]
@tomekit  Thanks for the workaround, it does look promising. However I can't find where to extract the query using the Chrome dev tools and am hence unable to copy the target parameter value for the new query. I have tried "Inspect Element" but I am struggling to find "Name" or "Headers" or "Form Data" that you have shown on the screenshot.

Would you please be able to illustrates the steps to do that? Your help will be highly appreciated!

Thanks

**Comment by tomekit** [comment]
@mathurj It is the Network -> XHR tab. Does it help now?
![image](https://cloud.githubusercontent.com/assets/7759113/25421508/8809a9e0-2a5d-11e7-8ac2-825c02627ff1.png)

**Comment by mathurj** [comment]
Thanks @tomekit I can see this page now however I cannot see any request named "render". There however is another request about the query that I am executing, but it does not have any "target" parameter. 

Any leads on how to go about the "render" request?

**Comment by tomekit** [comment]
@mathurj I get the "render" request, once looking on one of the graphs on my dashboard and clicking refresh (top right corner).

**Comment by mathurj** [comment]
Tried it, still no "render" request for me :( And no "target" paremeter too. Guess i'll have to wait for the actual implementation, which might be a while with the looks of it. @bergquist @torkelo ?

**Comment by SvenDhaens** [comment]
ok working with 
some.metric{hostname=~"$Hostname"}
in the query itself is fine,
but its my datasource that is the template here...
![prtscr_71](https://cloud.githubusercontent.com/assets/10166325/26211773/a031746a-3bf3-11e7-9fbd-4c78f9da2027.jpg)
environment=~"$environment"
doesn't seem to work... any way for this to work did I mis anything? or should I get rid of the template :disappointed:

**Comment by matthew-churcher** [comment]
+2

**Comment by cleberwarmling** [comment]
+1

**Comment by amalucelli** [comment]
+1

**Comment by dikang123** [comment]
this feature is especially useful when using the prometheus as the data source!

**Comment by drekle** [comment]
I need this as well, for similar reasons as mentioned above.  I expect this to work as a for each loop over the entire collection that the template defines.

**Comment by helletheone** [comment]
We need this!!! :) 👍

**Comment by Baiteman** [comment]
Personally I'm in support of this feature as well. We are testing more than one system with a set of load tests that generate delay metrics. Instead of having one dashboard we are using variable to switch between the datasources that hold the data for the various systems in order to just have to maintain one board and not script them.
Hence a support for templates in alerts would be highly appreciated.

**Comment by Eloo** [comment]
+1
we need this too

**Comment by svenkirsten** [comment]
+1
we need this too

**Comment by swift1911** [comment]
+1
we need this too

**Comment by skygragon** [comment]
Can I ask why there are so many guys "thumb down" those +1 replies??

**Comment by austinbutler** [comment]
@skygragon it's essentially useless spam when the option to +1 the original post exists. Just click the thumbs-up icon in the first post.

**Comment by secure12** [comment]
Template variables and alerting are 2 of the best features of Grafana.
Sad to see they are mutually exclusive though......

**Comment by marcomusso** [comment]
:+1:

**Comment by yzargari** [comment]
+1

**Comment by yesman85** [comment]
@bergquist could your team discuss this again and hopefully put a milestone on this? It's the most requested feature for almost 2 years now and I bet quite a few users would be happy with getting this feature.

**Comment by torkelo** [comment]
No good solution has yet been proposed and we are still pretty sure this is a bad idea as it's mixing two features that server different purposes. Templating are used for dynamic dashboards & exploration, alert rules can already by made to be dynamic with regex / wildcard queries. Mixing these to seems like a terrible idea, at least in way that is understandable & predictable. 

There are some good reasons though to support it in some limited way but not sure it's going to be worth the extreme complexity it would add and the development cost. But we are open to ideas, suggestions & PRs.

**Comment by yesman85** [comment]
The issue is I have quite a few servers I monitor AND they are dynamicaly created through Amazon, so at a given time I don't know how many servers are up.

I have a templated graph that shows CPU for each server (for example), so I would like alerting on there too.

But you are saying I could achieve the same by using wildcards?

**Comment by torkelo** [comment]
@yesman85 well of course, depends on your time series store. but most support some form of glob/regex query syntax to target multiple metrics that follow a naming pattern.

**Comment by pdf** [comment]
@torkelo I believe this is the first time that this position has been stated publicly.  I think maybe there's some misunderstanding here - I don't believe users want the template values selected for dashboard output to affect alerting, but instead to have the same template selections available when configuring alerts.

Relevant implementation suggestions from this thread:
https://github.com/grafana/grafana/issues/6557#issuecomment-272588490
https://github.com/grafana/grafana/issues/6557#issuecomment-281049641 (option 2)

Also, related issues:
https://github.com/grafana/grafana/issues/6041
https://github.com/grafana/grafana/issues/6553

These limitations have stopped us from being able to use Grafana for alerting in anger, because the only way to do so that sort of works currently is to add a bunch of separate hidden queries for alerting, which is a maintenance nightmare, and rather limits the great potential that Grafana has for easy monitoring configuration and interpretation by users.

**Comment by danhallin** [comment]
I think my suggestion above inspired by Libratos source-based alerting gives a limited, understandable and predictable solution which seem to cover almost all of the above mentioned problems.

In grafana it would translate to being able to use one and only one template variable for alerting and each value in that variable would generate its own alert. You could possible also add a regex on it for filtering in / out which ones you want to create alerts on.

**Comment by torkelo** [comment]
@pdf I agree that https://github.com/grafana/grafana/issues/6041 is a big limitation, one that we definitely want to fix but is unrelated to this issue. It's bad that we haven't fixed it yet I agree, we have been a bit understaffed on the alerting side last couple of months but that will change soon! 

@danhallin that is not the same, seems it would translate to a query in Grafana that targets many series using a wildcards or glob expressions in your query or only filters on a limited set of tags? Librato alert rules are defined separately from dashboards are they not so how can it translate to a dashboard wide  data filtering feature? 

> way to do so that sort of works currently is to add a bunch of separate hidden queries for alerting, which is a maintenance nightmare

Understand that mixing in alert rules in your templated graphs like that is a nightmare. But think supporting template variables in alerting could be an even bigger nightmare. Probably a stupid question, but can you not built dashboards & graphs that are focused on alerting? and leave the dashboards with many template variables for exploration and troubleshooting? I know there are probably many cases where they would be similar so feels like duplicated work :( The problem with alert rules in the context of a dashboard with a couple of template variables is how it should work & be understandable, and how it can even be implemented at all.  

Say one query uses 2 template variables, $A, and $B they each have 50 values. Would this result in evaluating the alert rule 2500 times? I mean if the variables are "single value", ie the queries are built to only work if $A and $B have a single value then we would have to do that. Not a show stopper maybe, we would have to explain only multi value variables are supported. There are probably many more limitations & problematic details that will make this feature very hard to implement & use & understand 

But I am not 100% against it, think there might be a way to do it in a limited way (like only supporting one 
 multi valued variable). There is also the use case of having multiple data sources and being able to target a datasource variable in your alert rule in order to reuse those alert rules across many data centers / environments (that have separate TSDBs) that user case would have to be solved using a data source template variable as a metric query with wildcards cannot solve it.

**Comment by pdf** [comment]
> but can you not built dashboards & graphs that are focused on alerting? and leave the dashboards with many template variables for exploration and troubleshooting? I know there are probably many cases where they would be similar so feels like duplicated work :(

I did actually edit my comment to reference this option (and add a couple of the other alert-related pain-point issues).  As you've identified though, this means multiplying dashboard/panel creation (and maintenance, if all those queries need an update) by the number of template variations you want to alert on, and also segregates the alert annotations from the location they may be most useful in - a templated dashboard with alert annotations can be very useful for making correlations, but not so much if you have to try and explore across multiple browser tabs and try to line things up.

> Say one query uses 2 template variables, $A, and $B they each have 50 values. Would this result in evaluating the alert rule 2500 times? I mean if the variables are "single value", ie the queries are built to only work if $A and $B have a single value then we would have to do that. Not a show stopper maybe, we would have to explain only multi value variables are supported. There are probably many more limitations & problematic details that will make this feature very hard to implement & use & understand

There are two separate issues here I think.  One _is_ (I believe) closely related to #6041, in that there may be a desire to evaluate alert conditions individually per series/template values (the aggregate toggle I mentioned in an earlier comment).  If we put that aside for now, I believe the ideal way to solve the bulk of this issue is to allow multiple alert configs per panel, and just do variable interpolation in exactly the same way as for dashboard output - allow users to select single- or multi-value template values when configuring alert queries; the queries will be executed once per alert config, with the selected values populated; and the results will be interpreted exactly the same way as they currently are.  Unless I'm grossly misunderstanding something, I don't see this as a significant increase in complexity, and should be quite user-friendly.

The ability to select template values to simply limit alerting scope would still be useful without multiple alerting configs (if it helps to develop this functionality in that order), but would be exponentially more valuable with multiple configs.

**Comment by vishwanathh** [comment]
This is definitely a popular need.. Now to achieve Alerting we had to move
away from Grafana and create our custom solution using Graphite's Render
APIs raw data ..  I don't think supporting Alerts in dynamic / templated
data is any complex atleast using Graphite..

One more thought is,. Why Grafana has Alerts section is part of dashboard
config if this is considered complex.  You can move it away to separate
Alerting view where users can jus input / config their dynamic query,
interval, evaluation condition over there..  May be this could mean we
don't hav duplicate dashboards.. Makes sense?



BR,
Vishwa..

**Comment by danhallin** [comment]
> @danhallin that is not the same, seems it would translate to a query in Grafana that targets many series using a wildcards or glob expressions in your query or only filters on a limited set of tags? Librato alert rules are defined separately from dashboards are they not so how can it translate to a dashboard wide data filtering feature?

I see now that what I'm discussing is basically a copy of: https://github.com/grafana/grafana/issues/6041

I'm in support of that feature request over this one.

**Comment by wavded** [comment]
We spin up servers regularly and we have templated dashboards that measure common things across all hosts (mem, hd, cpu, etc).  Making an "alert" dashboard having graphs for each possible combination is too tedious, many of the desired alerts can be generalized across all the hosts.  AFAIKT this issue is asking for a way to make this use case possible and would not be solved in #6041.  Unless I'm missing something.

**Comment by torkelo** [comment]
Making a graph & alert for every combination is not what we are proposing, just that you alert query uses wildcards or regex so that it covers all hosts etc. Then you have one alert and graph that can be tuned for alerting and not be competing with requirements for troubleshooting &  dynamic filtering.

**Comment by pdolan** [comment]
For our usecase we have a number of (potentially) ephemeral instances providing realtime processing of events - we'd need to be alerted if certain application generated metrics drop to 0. However, if instances are removed or halted they will still have a named metric associated with them that shows up in a straight wildcard search leading to falsely generated alerts.

Have gotten around this by using a templated, multi-value variable (set to All) generated from a "Simple JSON Datasource" that contains a source of truth for what instances *should* be running - all working perfectly. Alas, as soon as I attempted to enable alerting I was led to this feature request :)

Not sure how unique our particular use-case is but I'm not sure of any other way to alert on this without the use of templating - for now, we will need to continue alerting for these issues outside of Grafana (which is a shame - we're very heavy Grafana users!)

**Comment by olsib** [comment]
+1

**Comment by JonathanTroyer** [comment]
So one case where we don't have any of these issues is where the template variable is a constant type. For instance, we have multiple dashboards relying on a constant variable to limit the data on that dashboard to a particular resource (the reason we didn't use a multi-value variable is because each dashboard is different enough to justify different setups but close enough to justify a "template" dashboard). At least in this case (constant variables) nothing about the current alerting behavior needs to change.

**Comment by crazy-canux** [comment]
Any news about this topic?

**Comment by ashuw018** [comment]
Hi,

Is there any hopes of getting this feature? i just wonder how other systems are having this features as they must also be using sort of templating as when we install agent it automatically gets appears on portal and alerting can also be set for that. (This is my experience with New Relic).

@vishwanathh  i liked the approach of having separate section for alerting (if it is being complicated to has it in graph panel) where we can put in our queries just for alerting. as this way our users will not see the placeholder panel(used for alerting).

Sorry for the extra noise but this would be a really great feature to have in Grafana.

**Comment by tangyong** [comment]
+1, very very important feature!

**Comment by tangyong** [comment]
In addition, if letting me modify prometheus metric query expression to remove template variable, this is not feasible at all. So, I think that this feature is most important for the prometheus+grafana to land on production!

Anyway, please team can consider the priority, thanks!

**Comment by pdf** [comment]
With 5.0 heading out the door shortly, I'd love to see some significant focus given alerting during the next release series.  Looking at Github reactions, alerting-related deficiencies appear to have far-and-away the most interest from users.

I know there has been some reluctance to tackle these things due to UI/UX complexity concerns, however I'm not convinced these concerns are necessarily justified.  Is there anything we as users can do that might help planning/design or to move these issues forward, short of pull-requests with actual code?

**Comment by ashuw018** [comment]
@torkelo This has helped me to setup alerting for all of my hosts using tags and now my each alerting graph contains multiple series formed by the combination of tags. Everything seemed to be working fine. But going through the docs and other issues i realized that if any of the series within graph has already took alerting state then alerts for other series will not be trigger if they also crossing limit.

Thats again being limitations.

Thanks.

**Comment by deiv061** [comment]
Any news about this feature ?

**Comment by spiffytech** [comment]
What's the effort for an outsider to contribute this feature?

**Comment by sjayaraman** [comment]
+1

**Comment by nookalavikas** [comment]
+1

**Comment by amihura** [comment]
+1

**Comment by Moon-Tae-Kwon** [comment]
We hope to be resolved.

**Comment by razvip91** [comment]
+1

**Comment by calebtote** [comment]
I don't want to beat a dead horse here, but we're having the same issue, and I want to provide some context as to why the existing proposals don't work in all circumstances. I also have a couple of ideas for workarounds, but why we need _some features_ to help make the workarounds sufficient. 

For all scenarios below, we're using a single templated variable:  `$env`

### Why don't you have an alerting dashboard?
We want to alert on a couple different environments, not just production. So we'd now need to have the same metric in _at least_ 3 different places (the troubleshooting dashboard with all metrics, not just the metric we alert for; the prod alerts dashboard; the integration alerts dashboard). This can get out of hand pretty quick, and is prone to user error.

Equally as important, this nullifies much of the gain from automated annotations from alerts. If I have to go back and forth from my exploratory dashboard to my alerts dashboard to see the annotations for when an event started and when it ended, that's going to be pretty tedious. 

### Attempted Solution
What we've done to try to get around this is we've added duplicate metrics _specifically for alerting_ to our dashboards. So if there's a metric we want to alert on, we go to the panel and add explicit metrics for those alerts (and hide them).

Our series list for a given panel that needs to alert will look along the lines of:
```
$env.myApp.getStatus.200
prod.myApp.getStatus.200
int.myApp.getStatus.200
```
With the non-templated series marked as hidden. Then in the alerting tab, we set thresholds for _these_ series, not the variable series.

![screen shot 2018-04-05 at 4 40 19 pm](https://user-images.githubusercontent.com/1459202/38390712-1c0ad0be-38f0-11e8-84bd-bb021ed827f3.png)

### Attempted Solution - Problems
This doesn't work great though. For example:
![screen shot 2018-04-05 at 4 43 21 pm](https://user-images.githubusercontent.com/1459202/38390834-7e2466d4-38f0-11e8-8150-cf6d98103af9.png)

![screen shot 2018-04-05 at 4 42 41 pm](https://user-images.githubusercontent.com/1459202/38390808-67d80890-38f0-11e8-9f7f-d7ef8436f237.png)

As you can see, the **Alerts** panel doesn't allow us to specify _which environment_ is alerting -- so we have to drill down into the alert to figure out which environment is borked at the moment. However, an **easy fix** for this might be just allowing the description to be as verbose as the **Alert History** panel that shows state transitions:

![screen shot 2018-04-05 at 4 44 33 pm](https://user-images.githubusercontent.com/1459202/38390902-ae57f3b6-38f0-11e8-9551-dd400cdf892c.png)

This is at least somewhat helpful, but even in this panel there's no indication of which alert has gone back to `Healthy` (the description from the above screenshot was derived from the alias we set on the series if anyone is wondering how to at least get that much to show up).

### Things that would help until this specific ticket is resolved
- Allow the option for displaying the series alias instead of a typed description for the alert (this way the alias can at least specify the $variable it's alerting for)
- Allow the state transition back to healthy to also show the series alias (in the History screenshot above)
- Allow a legend value for active alerts (using the series alias I assume) for a given panel

**Comment by PheonixS** [comment]
+1

**Comment by stereosonicstar** [comment]
+1

**Comment by unixway** [comment]
+1

**Comment by PheonixS** [comment]
plus one :)

**Comment by adamcstephens** [comment]
Please stop +1'ing this issue. It generates unnecessary spam emails. The ability to add a reaction to a github issue comment has existed for a while now, and over 429 people have figured out how to like the initial comment instead of spamming everybody who is subscribed.

**Comment by manueligno78** [comment]
Please we really need this feature, we would like to use templating, but in our case is most important to have a clear alerting system. So to workaround this we are avoiding templating in our dashboard... its a mess.

**Comment by napsy** [comment]
+1

**Comment by napsy** [comment]
+1

**Comment by thiagocorredor** [comment]
I agree and this feature will help us a lot !!!!

**Comment by kamzhuyuqing** [comment]
+1 please

**Comment by cmuzyunda** [comment]
we need this

**Comment by ChahatB** [comment]
+1 please. It's really needed.

**Comment by pdf** [comment]
@bergquist @torkelo can we please lock this issue to stop the +1 spam?

**Comment by jessover9000** [comment]
Hello Grafana community, the Grafana team has picked up the work on Alerting and we're in the process of redesigning it to make the best possible alerting experience happen 🔥 🚀  We would love to find out more about your needs as our beloved users. So if any of you are willing to have a 30-minute interview with me, please just [send me an empty e-mail](mailto:jessica.muller@grafana.com?subject=Alerting%20research%20interview) and I will get in touch. Looking forward to hearing from you! :)

**Comment by kylebrandt** [comment]
The new beta version of alerting in Grafana 8 ([opt-in with "ngalert" feature toggle]( https://grafana.com/docs/grafana/latest/alerting/unified-alerting/opt-in/)) has moved alerts out of dashboards, so alert rule queries based on template variables no longer fits into the model.

The new version does support "multi-dimensional" alerting based on labels which seems to be close to the core functionality underlying this issue. So one can have multiple alert instances from a single rule. More at https://github.com/grafana/grafana/issues/7832#issuecomment-856725093 .

**Comment by fkaleo** [comment]
> The new beta version of alerting in Grafana 8 ([opt-in with "ngalert" feature toggle](https://grafana.com/docs/grafana/latest/alerting/unified-alerting/opt-in/)) has moved alerts out of dashboards, so alert rule queries based on template variables no longer fits into the model.
> 
> The new version does support "multi-dimensional" alerting based on labels which seems to be close to the core functionality underlying this issue. So one can have multiple alert instances from a single rule. More at [#7832 (comment)](https://github.com/grafana/grafana/issues/7832#issuecomment-856725093) .

Is there a way to cover this use case:
As a user, if I see a metric in a dashboard with template variables that I would like to create an alert for, is it possible to easily create that alert (in the ngalert model) that is based on that query (with the template variables replaced by their current values)

**Comment by kylebrandt** [comment]
@fkaleo It doesn't seem to currently do this. The resolution of those variables when clicking the "create alert" from the dashboard in the system would be on the frontend side which I know less about it. But it sounds doable and like a good idea to me - can you create a new enhancement request for that?

**Comment by UAnton** [comment]
Can't find ngalert feature toggle

**Comment by kylebrandt** [comment]
> Can't find ngalert feature toggle

If using `.ini` files for your configuration, it would be adding the following to the config:

```ini
[feature_toggles]
enable = ngalert
```

**Comment by anybodysguest** [comment]
@torkelo I've been keeping up to date with the latest Grafana releases hoping that eventually I'd be able to use variables in alert queries. I tried ngalerts and while I didn't spend a lot of time with them I had the following thoughts:

1. When this feature is turned on, existing alerts become invalid.  Non-ngalerts don't co-exist with the newer ngalerts.
2. I found that ngalerts was overly complicated, at least compared to the non-ngalerts version.
3. While ngalerts may bring the ability to use template variables I didn't test it due to the fact I was turned off by the fact that all of my existing alerts would effectively have to be re-written.
4. I use my alerts with a web hook to xMatters. Previously I was able to add tags to my alerts and those tags would be sent to xMatters.  My tags specify the type of alert is being fired (for instance CANARY) and the severity(example: High).  When the xMatters webhook receive that webhook, I can "unpack" those tags to determine that the flow my incident workflow will take.  With ngalerts this no longer works because the format of the payload being sent to xMatters has changed.  There is no longer a Tags array in the payload.

So for net-new dashboards and alerting ngalerts probably is a great thing.  But what happens to many of us who are not using ngalerts and simply want to use a contant template variable in the alert query or in the alarm message itself?

I still don't understand why template variables can't be used with alert queries.  I use template variables heavily, whenever I can so that I have precise control over my dashboard queries.  In that respect there is no problem. It is only when I have an alert attached that things go sideways.  That is an "impedance-mismatch" I don't understand.

**Comment by Guanpeng520** [comment]
Hi all, had a question about adding alerts.
[https://grafana.com/docs/grafana/next/alerting/old-alerting/add-notification-template/]( look as  document)

 it’s suggested that the query triggering the alert can have labels from the query templated into the alert message. As per the documentation, this is how:

Refer to the alert query labels in the alert rule name and/or alert notification message field by using the ${Label} syntax.

![Uploading image.png…]()

Who knows where the ${Label} was created? Thanks so much.

**Comment by lujinke** [comment]
Yes, I don't understand why the Grafana dev team does not want to support this while so many many people are eager for this.

**Comment by DEvil0000** [comment]
someone may think because their business model is to make the software just good enough and make people pay for support to code the other little things missing.
did someone try to PR this? or pay for it?

22.03.2022 11:38:42 lujinke ***@***.***>:

**Comment by Tomasz-Kluczkowski** [comment]
Any progress on this Grafana? Sorry to be a nuisance but this pretty much invalidates the template variables for me.
If I use a variable to define graphs but I cannot use it for alert and need to re-do the query without a variable, what is the point?

**Comment by amohamedhey** [comment]
Hey, Any update on this?

**Comment by svet-b** [comment]
It's not quite clear what sort of further updates people are expecting on this. This feature **has been implemented** (per [comment above](https://github.com/grafana/grafana/issues/6557#issuecomment-856758319)) and the issue is correspondingly closed.

Additional comments (and I'm conscious I'm right now a culprit here...) simply generate spam for the 800+ people still subscribed to the issue. Follow-up bug reports or enhancement requests ought to be captured in separate [issues](https://github.com/grafana/grafana/issues), while questions about the functionality can be addressed in the [community forum](https://community.grafana.com).

**Comment by anybodysguest** [comment]
@amohamedhey @svet-b @Tomasz-Kluczkowski My understanding is that this ability was added to Grafana for Unified Alerting (introduced in Grafana 8).  I don't think a lot of new feature work will be going into the legacy alerts as Unified Alerting is the "new normal".  That is my understanding. @svet-b please correct me if I'm wrong.
So, that said, if you are using legacy alerting, you still won't be able to use template variables in the alert queries.  If you do you will be warned, if you proceed to use them they will disable the alert.   If you have the ability to easily switch to Unified Alerting, I'd do that.  I don't think Legacy and Unified Alerting can be used at the same time on a Grafana installation; I haven't checked to see if that's possible.

**Comment by dylan-tao** [comment]
Hey, Any update on this?
